### DISCUSSION TITLE
Interesting Results for Flexible Job Shop Problem

### DISCUSSION THREAD

**Discussion by hpswalters** [google/or-tools#2614] [CP-SAT questions] [answered]
First, thanks to the OR-Tools Team.  This is one of the most valuable set of free tools I've found in the last several years.

I've been running the flexible job shop (flexible flow shop, really), modifying code found published through this site, on a problem with 198 jobs each with four available routings (machines/alternatives).  Different routings have different runtimes and set-up times.  I know that this set of jobs can be completed in less than 7200 minutes using all four lines and this scheduling is currently done manually. .  When run as a single program (all 198 jobs across 4 lines) using CP SAT, the results really don't come close to that (running for 3 hrs).  However, when I run this iteratively, one line at a time, starting with the fastest line and all 198 jobs, find a solution, and then remove all jobs beyond the 7200 minute time limit and running again on the next fastest line (and so on), the results are wonderful--The first two lines are used to capacity (7200), the third less than half ( I can cut a shift) and the fourth is not required at all.  Since this requires some manual manipulation on my part, there are breaks between runs but the total runtime for the solution is 20 minutes (run 1), 5 minutes (run 2) and less than one minute, solved to optimality (run 3)

What started me down this track was an analysis of just how long it took to do all 198 jobs on the fastest line.  This came in at a surprising 12.5K minutes.  12.5K minutes on one line and 8K minutes on four lines.  That just didn't seem right.  Then I ran all 198 jobs using just two lines.  The results came in at about 10K minutes and that just didn't seem right.

Any insights, other than problem complexity, that may yield some insights as to the extreme difference in solution quality and performance.  Program structure? Objective function? I've run this on a variety of machines AMD 5950x (16 threads), 3970x(32 threads), i9-9900k (16 threads) and while results vary the single run, 4 lines, comes in at the high 7k to low 8k minute range. 

I can publish the code and data if anyone is interested.

Thanks,
hpsw

**Comment by gregy4** [comment]
Maybe some kind of error in your code or too complex model, I solve scheduling problems with much more jobs (thousands) and alternatives (hundreds).
Please publish the code and data.

**Comment by hpswalters** [comment]
Here you go.  Let me know if you can get to it.
Within the python code, I've provided some hints that come from a superior solution.  I doubt this problem is as complex as yours so I'm pretty sure the issues are related to problem structure

Any insights would be appreciated

Best,
Hugh


[flexible flow shop scheduling w setup duedates, releasedates large.txt](https://github.com/google/or-tools/files/6856350/flexible.flow.shop.scheduling.w.setup.duedates.releasedates.large.txt)

**Reply by lperron** [reply]
Maybe the model is wrong. Can you inject a manual solution ?

**Reply by hpswalters** [reply]
Thanks for taking a look your expertise is appreciated.  I don't have a digital copy of the manual plan as it was in document format--not digital.  However, I do have the solution to the successive runs which I have attached
[Job Shop Solution.xlsx](https://github.com/google/or-tools/files/6868235/Job.Shop.Solution.xlsx)

**Comment by lperron** [comment, marked as answer]
A few ideas:
-  I removed everything around ranks (variables, containers, constraints). They are not use and just slow the solver down.
- You can add start_lit implies task is performed in the circuit constraint
- I removed the various constraints on min/max machine capacity. They make the LNS much harder.

On my workstation, I get a solution of 7693 in a bit more than 15 minutes (using 32 workers).

Starting Search at 5.04s with 32 workers and subsolvers: [ default_lp, reduced_costs, pseudo_costs, no_lp, max_lp, quick_restart, quick_restart_no_lp, probing, lb_tree_search, feasibility_pump, rnd_var_lns_default, rnd_cst_lns_default, graph_var_lns_default, graph_cst_lns_default, scheduling_time_window_lns_default, scheduling_random_lns_default, rins_lns_default, rens_lns_default ]
#1       9.26s best:20087 next:[0,20086]  no_lp fixed_bools:1/161371
#2       9.59s best:20046 next:[0,20045]  no_lp fixed_bools:1/161410
#3       9.73s best:20045 next:[0,20044]  no_lp fixed_bools:1/161410
#4       9.84s best:20039 next:[0,20038]  no_lp fixed_bools:1/161604
#5       9.89s best:12367 next:[0,12366]  quick_restart_no_lp fixed_bools:1/161371
#Bound  13.94s best:12367 next:[197,12366] lb_tree_search
#Bound  14.25s best:12367 next:[331,12366] lb_tree_search
#Bound  14.44s best:12367 next:[414,12366] lb_tree_search
#Bound  14.75s best:12367 next:[453,12366] lb_tree_search
#Bound  15.09s best:12367 next:[462,12366] lb_tree_search
#6      15.46s best:12339 next:[462,12338] quick_restart_no_lp fixed_bools:1/166209
#7      17.72s best:12338 next:[462,12337] quick_restart_no_lp fixed_bools:1/166997
#8      18.03s best:12316 next:[462,12315] quick_restart_no_lp fixed_bools:1/167714
#9      21.82s best:12311 next:[462,12310] quick_restart_no_lp fixed_bools:1/174652
#10     21.94s best:12309 next:[462,12308] quick_restart_no_lp fixed_bools:1/175264
#11     25.63s best:12287 next:[462,12286] quick_restart_no_lp fixed_bools:1/176651
#12     26.95s best:10699 next:[462,10698] rins_lns_default_lp(d=0.50 s=297 t=0.10 p=0.00)
#13     29.16s best:10688 next:[462,10687] quick_restart_no_lp fixed_bools:1/178037
#14     30.49s best:10659 next:[462,10658] quick_restart_no_lp fixed_bools:1/178597
#15     31.10s best:10649 next:[462,10648] quick_restart_no_lp fixed_bools:1/179242
#16     33.95s best:9188  next:[462,9187] rins_lns_default_lp(d=0.71 s=430 t=0.10 p=1.00)
#17    103.55s best:8910  next:[462,8909] quick_restart_no_lp fixed_bools:1/185100
#18    105.11s best:8907  next:[462,8906] quick_restart_no_lp fixed_bools:1/197958
#19    108.08s best:8872  next:[462,8871] rins_lns_default_lp(d=0.99 s=2122 t=0.10 p=1.00)
#20    114.77s best:8871  next:[462,8870] quick_restart_no_lp fixed_bools:1/198323
#21    131.87s best:8363  next:[462,8362] rins_lns_default_lp(d=1.00 s=2554 t=0.10 p=1.00)
#22    198.68s best:8184  next:[462,8183] quick_restart_no_lp fixed_bools:1/205022
#23    230.96s best:8113  next:[462,8112] quick_restart_no_lp fixed_bools:1/207038
#24    329.04s best:8106  next:[462,8105] rins_lns_default_lp(d=1.00 s=6901 t=0.10 p=0.98)
#25    499.45s best:8039  next:[462,8038] quick_restart_no_lp fixed_bools:1/238584
#26    514.48s best:7964  next:[462,7963] graph_var_lns_default(d=0.36 s=11082 t=0.10 p=0.44) [hint]
#27    518.73s best:7924  next:[462,7923] graph_var_lns_default(d=0.30 s=11163 t=0.10 p=0.42)
#28    595.86s best:7868  next:[462,7867] quick_restart_no_lp fixed_bools:1/257193
#29    921.45s best:7854  next:[462,7853] rins_lns_default_lp(d=1.00 s=18790 t=0.10 p=0.97)
#30    1052.07s best:7792  next:[462,7791] quick_restart_no_lp fixed_bools:1/289506
#31    1053.17s best:7694  next:[462,7693] rins_lns_default_lp(d=1.00 s=21292 t=0.10 p=0.96)

**Reply by hpswalters** [reply]
Thanks for taking the time.  Your results are Interesting.  Your run closes to a better solution faster and, I assume, with greater regularity.  I'd be curious what a longer run reveals.  Would you publish your code modifications so that I can make a longer run?

Thanks,
Hugh

**Reply by lperron** [reply]
[issue2614.py.txt](https://github.com/google/or-tools/files/7698273/issue2614.py.txt)

**Reply by lperron** [reply]
Finally managed to upload the file.

**Reply by lperron** [reply]
with good results (my laptop, 16 workers)

```
#30    301.68s best:7447  next:[462,7446] routing_random_lns_default(d=0.15 s=269 t=0.10 p=0.48)
Solution 29, time = 301.750273 s, objective = 7447, makespan = 7447
```

**Reply by lperron** [reply]
Note the routing (circuit) based LNS does wonder here.

**Reply by hpswalters** [reply]
Laurent,
Thanks for taking the time to do this.  The changes seem so subtle.  Looking forward to running then

**Comment by lperron** [comment]
Here is the code.


Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00



Le ven. 23 juil. 2021 à 13:15, hpswalters ***@***.***> a
écrit :

**Reply by hpswalters** [reply]
Sorry Laurent,  Not seeing a link to your code

**Comment by lperron** [comment]
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00



Le ven. 23 juil. 2021 à 19:15, hpswalters ***@***.***> a
écrit :

**Comment by hpswalters** [comment]
Laurent,
Thanks so much for this.  I'm not sure what had a greater impact--changes to the code or changes to or-tools.
I ran your code with version 8.1.8487 and there really was a small improvement.  I have since installed the latest version of or-tools and the improvement is dramatic  After 3 minutes on an AMD 3970 (32 cores, 32 workers), the solution is below the 7200-minute threshold.  After 20 minutes it is below 6000 minutes.  The number of workers only seems to have a minor impact on solution times as nearly the same results are generated with 16 cores, 16 workers (AMD 5950) and 8 cores, 8 workers (AMD 5900HX laptop)

When I have some time, I will run the old code with the new version of or-tools and post the results here.  

Thank you so much for you help
-hpsw

**Reply by hpswalters** [reply]
Updating to the new version of CP-SAT has had the most significant impact on the solution.  I've run the old code on the new version of CP-SAT and got very comparable results

**Comment by lperron** [comment]
The biggest improvement is the routing_lns
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00



Le mer. 5 janv. 2022 à 19:14, hpswalters ***@***.***> a
écrit :

**Reply by hpswalters** [reply]
You must have done some real magic there since the solution quality and speed are much improved.  Thank you again
